### ISSUE TITLE
About the train.py and predict.py

### ISSUE THREAD

**Issue by rlcgetUp** [GeoZcx/A-deeply-supervised-image-fusion-network-for-change-detection-in-remote-sensing-images#26]
Can you share the  trianing.py and predict.py of troch version. I will appreciate you very much.

**Comment by rlcgetUp** [comment]
尊敬的论文作者您好你好，我是计算机研二学生。在复现代码时您提供的tensorfolw版本的代码不全，自己无法看懂。希望您可以通过github上传或者百度网盘的方式分享一下pytorch版本的训练代码。因为我将pytorch版本的模型用到自己的训练文件上时图中红框表示的model_A,和model_B不知道该传递什么参数，所以我目前无法实现模型的复现，希望作者可以分享您完整的代码或者是告诉一下我model_A,model_B应该怎么理解，传递什么参数，我将十分感激。







------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Chenxiao ***@***.***&gt;; 
发送时间: 2022年4月11日(星期一) 上午10:32
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [GeoZcx/A-deeply-supervised-image-fusion-network-for-change-detection-in-remote-sensing-images] About the train.py and predict.py (Issue #26)





 
Closed #26.

**Comment by songyvying** [comment]
您好，因为我最近在复现这篇论文的代码时，发现代码不全，因为刚刚接触这个方向，希望获得完整的代码来学习，如果您获得了完整的代码，是否可以分享一下呢，我的邮箱是1941167521@qq.com,万分感谢！

**Comment by rlcgetUp** [comment]
你好，我的代码也不能训练，我现在训练了40轮了精度都很低，我也不知道什么原因。



------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2022年4月29日(星期五) 晚上8:08
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [GeoZcx/A-deeply-supervised-image-fusion-network-for-change-detection-in-remote-sensing-images] About the train.py and predict.py (Issue #26)





 
***@***.***,万分感谢！

**Comment by songyvying** [comment]
你好！我刚刚在邮箱回复您了，不知道您有没有看到，本菜鸟真的想先复现一下，先不考虑精度的问题，所以如果您有预测部分的代码，是否介意分享一下呢？再次对您表示感谢！！！！邮箱1941167521@qq.com

**Comment by GeoZcx** [comment]
> 尊敬的论文作者您好你好，我是计算机研二学生。在复现代码时您提供的tensorfolw版本的代码不全，自己无法看懂。希望您可以通过github上传或者百度网盘的方式分享一下pytorch版本的训练代码。因为我将pytorch版本的模型用到自己的训练文件上时图中红框表示的model_A,和model_B不知道该传递什么参数，所以我目前无法实现模型的复现，希望作者可以分享您完整的代码或者是告诉一下我model_A,model_B应该怎么理解，传递什么参数，我将十分感激。
> […](#)
> ------------------&nbsp;原始邮件&nbsp;------------------ 发件人: "Chenxiao ***@***.***&gt;; 发送时间: 2022年4月11日(星期一) 上午10:32 收件人: ***@***.***&gt;; 抄送: ***@***.***&gt;; ***@***.***&gt;; 主题: Re: [GeoZcx/A-deeply-supervised-image-fusion-network-for-change-detection-in-remote-sensing-images] About the train.py and predict.py (Issue #26) Closed #26. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you authored the thread.Message ID: ***@***.***&gt;

[你好！非常抱歉我刚刚看到关闭的问题。训练部分的代码我贴在这里供你参考，目前仅有keras版本的，pytorch版本我仅复现了模型部分，train和predict的代码没有复现，git上有很多开源的train代码，都可以快速复现。对于训练不收敛的问题，建议修改下学习率和优化器。

from keras.models import model_from_json,load_model
import os
import numpy as np
import rasterio
from losses import dice_loss,balanced_loss,bce_dice_loss
from metrics import f1
from keras.preprocessing import image

os.environ["CUDA_VISIBLE_DEVICES"] = '2'


def read_4bands_image(file_path):
    ras_dataset = rasterio.open(file_path)
    ras_band1 = ras_dataset.read(1)
    ras_band2 = ras_dataset.read(2)
    ras_band3 = ras_dataset.read(3)

    band_stack = np.array([ras_band1, ras_band2, ras_band3])
    band_stack = np.transpose(band_stack, (1, 2, 0))
    list_tc_0 = np.array(band_stack)

    return list_tc_0

def predict():
    train_t1_path = os.path.join("/home/zcx/ChangeDetectionDataset/test","t1")
    train_t2_path = os.path.join("/home/zcx/ChangeDetectionDataset/test","t2")

    model = load_model('best_model.h5', custom_objects={'bce_dice_loss': bce_dice_loss,'f1':f1})


    list_dir = os.listdir(train_t1_path)

    for i in list_dir:
        t1_data = []
        t2_data = []
        train_t1 = read_4bands_image(os.path.join(train_t1_path,i))/255.0
        train_t2 = read_4bands_image(os.path.join(train_t2_path,i.replace("t1","t2")))/255.0
        t1_data.append(train_t1)
        t2_data.append(train_t2)
        t1_data = np.array(t1_data)
        t2_data = np.array(t2_data)
        # train_change = np.concatenate([t1_data,t2_data],axis=3)

        result = model.predict([t1_data,t2_data])[3]
        predict_path = os.path.join("predicted", i)
        result = result[0,:,:,:]

        result = np.transpose(result, (2,0,1))
        result = result[0]
        result[result<0.5] = 0.0
        result[result >= 0.5] = 255.0

        #save image
        bands_combination = np.zeros(shape=(256,256,3))
        bands_combination[:, :, 0] = result
        bands_combination[:, :, 1] = result
        bands_combination[:, :, 2] = result

        predict_result = image.array_to_img(bands_combination)
        image.save_img(os.path.join('predicted', i), predict_result)

predict()

**Comment by rlcgetUp** [comment]
十分感谢您的回复，我用pytorch的训练代码跑了pytorch的模型，但是精度上不去，我也不知道为什么，然后就谢谢您的回复了，我再研究研究。



------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2022年5月31日(星期二) 晚上9:01
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [GeoZcx/A-deeply-supervised-image-fusion-network-for-change-detection-in-remote-sensing-images] About the train.py and predict.py (Issue #26)





  
尊敬的论文作者您好你好，我是计算机研二学生。在复现代码时您提供的tensorfolw版本的代码不全，自己无法看懂。希望您可以通过github上传或者百度网盘的方式分享一下pytorch版本的训练代码。因为我将pytorch版本的模型用到自己的训练文件上时图中红框表示的model_A,和model_B不知道该传递什么参数，所以我目前无法实现模型的复现，希望作者可以分享您完整的代码或者是告诉一下我model_A,model_B应该怎么理解，传递什么参数，我将十分感激。
 …
 ------------------&nbsp;原始邮件&nbsp;------------------ 发件人: "Chenxiao @.&gt;; 发送时间: 2022年4月11日(星期一) 上午10:32 收件人: @.&gt;; 抄送: @.&gt;; @.&gt;; 主题: Re: [GeoZcx/A-deeply-supervised-image-fusion-network-for-change-detection-in-remote-sensing-images] About the train.py and predict.py (Issue #26) Closed #26. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you authored the thread.Message ID: @.***&gt;
  
[你好！非常抱歉我刚刚看到关闭的问题。训练部分的代码我贴在这里供你参考，目前仅有keras版本的，pytorch版本我仅复现了模型部分，train和predict的代码没有复现，git上有很多开源的train代码，都可以快速复现。对于训练不收敛的问题，建议修改下学习率和优化器。
 
from keras.models import model_from_json,load_model
 import os
 import numpy as np
 import rasterio
 from losses import dice_loss,balanced_loss,bce_dice_loss
 from metrics import f1
 from keras.preprocessing import image
 
os.environ["CUDA_VISIBLE_DEVICES"] = '2'
 
def read_4bands_image(file_path):
 ras_dataset = rasterio.open(file_path)
 ras_band1 = ras_dataset.read(1)
 ras_band2 = ras_dataset.read(2)
 ras_band3 = ras_dataset.read(3)
 band_stack = np.array([ras_band1, ras_band2, ras_band3]) band_stack = np.transpose(band_stack, (1, 2, 0)) list_tc_0 = np.array(band_stack) return list_tc_0  
def predict():
 train_t1_path = os.path.join("/home/zcx/ChangeDetectionDataset/test","t1")
 train_t2_path = os.path.join("/home/zcx/ChangeDetectionDataset/test","t2")
 model = load_model('best_model.h5', custom_objects={'bce_dice_loss': bce_dice_loss,'f1':f1}) list_dir = os.listdir(train_t1_path) for i in list_dir:     t1_data = []     t2_data = []     train_t1 = read_4bands_image(os.path.join(train_t1_path,i))/255.0     train_t2 = read_4bands_image(os.path.join(train_t2_path,i.replace("t1","t2")))/255.0     t1_data.append(train_t1)     t2_data.append(train_t2)     t1_data = np.array(t1_data)     t2_data = np.array(t2_data)     # train_change = np.concatenate([t1_data,t2_data],axis=3)     result = model.predict([t1_data,t2_data])[3]     predict_path = os.path.join("predicted", i)     result = result[0,:,:,:]     result = np.transpose(result, (2,0,1))     result = result[0]     result[result<0.5] = 0.0     result[result &gt;= 0.5] = 255.0     #save image     bands_combination = np.zeros(shape=(256,256,3))     bands_combination[:, :, 0] = result     bands_combination[:, :, 1] = result     bands_combination[:, :, 2] = result     predict_result = image.array_to_img(bands_combination)     image.save_img(os.path.join('predicted', i), predict_result)  
predict()
